certbot/certbot/_internal/client.py

```diff
@@ -454,8 +457,11 @@ def obtain_certificate(self, domains: List[str], old_keypath: Optional[str] = No
                 os.remove(csr.file)
             return self.obtain_certificate(successful_domains)
```

Comment:
Right now we have some logging for retries originating from `newOrder` and `finalize`.

That leaves the middle case, where some of the authorizations have failed, and Certbot retries without those domains.

I didn't realize that we had no logging at all here. Certbot currently drops the failed domains silently.

Whatever we decide to do, I think it should be consistent across all 3 cases.

Also, after sleeping on it, perhaps `", ".join(failed_domains)` is better than the abbreviated summary list. For purposes of later identifying "what happened", it would be good if the log file contained the comprehensive list of dropped domains.

Reply:
Switched from `summarize_domain_list` to your suggestion.

Are you suggesting that we add a logger call to `_retry_obtain_certificate`, or elsewhere? 

I think a debug or warning call in there makes sense to surface the failed domains (and maybe the successful domains we're retrying with?) in the logfile. Do you think it's also worth a `logger.debug` above each _call_ to `_retry_obtain_certificate` too, so we give some context as to where the retry came from?

Reply:
I am talking about the third place where we retry.

We have two code paths which retry via `_retry_obtain_certificate`, and those are triggered by errors from `newOrder` and `finalize`.

But we have a third code path which retries, triggered by one or more authorizations failing. But does not call `_retry_obtain_certificate`, so there is no logging:

https://github.com/certbot/certbot/blob/8c9b3027534ea65cd2f5f5845f153e35950d2a1c/certbot/certbot/_internal/client.py#L445-L458

There should be logging here, and it should be identical to the logging for the other two retry paths.

Alternatively, you could try to consolidate all of the retry logic into `_retry_obtain_certificate`, so that it will handle both cases:

- An ACME error with subproblems with identifiers, *or*
- One or more of the authorizations failed

Reply:
Thanks, I think I understand now.

I pulled the logic that builds `successful_domains` out of `_retry_obtain_certificate`, as that is the main behaviour difference between retrying based on subproblems with identifiers vs when one or more authorizations failed.

`_retry_obtain_certificate` now takes a list of `successful_domains` rather than `error` and is used in both of the aforementioned cases.